twister: allow building a platform with multiple toolchains (#114093)
zephyrproject-rtos/zephyr: doc/develop/twister/index.rst

toolchain:
  The list of supported toolchains that can build this board. This should match
  one of the values used for :envvar:`ZEPHYR_TOOLCHAIN_VARIANT` when building on the command line

> build_toolchains:

Reviewer: 1. Can you please document how will this relate to --force-toolchain , the preferred-toolchain key, and the environment set ZEPHYR_TOOLCHAIN_VARIANT. (maybe there should be a separate subsection referenced from all these options explaining how they all relate)

2. (As a bonus) Can you please document `preferred-toolchain`? (it is missing now)

  for that specific variant, while inheriting the remaining values from the
  top-level definition.

.. _twister_toolchain_selection:

Toolchain Selection
*******************

Several options influence which toolchain a test is built with. They fall into
three groups: options that *select* a toolchain, options that *filter* out test

> instances whose toolchain is not usable, and options that *multiply* a test into
> several builds.

Reviewer: ```suggestion
instances whose toolchain is not usable, and options that *replicate* a test
into an instance per toolchain.
```
or
```suggestion
instances whose toolchain is not usable, and options that *instantiate* a test per toolchain.
```

Nit: This text uses "build" in a few places , but the test will be build and run with each toolchain. So maybe this proposed phrasing is a bit clearer.

Twister first determines a default toolchain for the whole run by invoking
``cmake/verify-toolchain.cmake``, which honors the :envvar:`ZEPHYR_TOOLCHAIN_VARIANT`
environment variable. This value is reported at the start of the run as
``Using '<toolchain>' toolchain variant.``

For every test scenario and platform pair, the toolchain is then selected using
the first of the following that applies:

> #. The test scenario's ``integration_toolchains``, if set. The test is built once
>    per listed toolchain.

Reviewer: ```suggestion
#. The test scenario's ``integration_toolchains``, if set. A test instance is
spawned per listed toolchain.
```
maybe better?

#. The platform's ``build_toolchains``, if set, either in the board configuration
   or in the :ref:`Twister configuration file <twister_test_config>`. The test is

>    built once per listed toolchain.

Reviewer: ```suggestion
   instantiated once per listed toolchain.
```
?
